File: backend/src/middleware/verifyToken.ts
import { NextFunction, Request, Response } from 'express'
import { prisma } from '../client';

const verifyToken = async (req: Request, res: Response, next: NextFunction) => {
  // we get the header from the request
  const authHeader = req.headers.authorization;
  
  if (!authHeader) {
    return res.status(401).json({ message: 'Error: no token provided' });
  }
  
  // we get the token from the header
  const token = authHeader.split(' ')[1];

  if (!token) {
    return res.status(400).json({ message: 'Error: token not found' });
  }

  // find the user with the token
  try {
    const user = await prisma.user.findUnique({
      where: {
        token: token as string,
      },
      select: {
        id: true,
        username: true,
      },
    });

    if (!user) {
      return res.status(401).json({ message: 'Invalid token' });
    }

    // set the user in the request
    (req as any).user = user;

    next();
  } catch (error) {
    res.status(500).json({ message: 'Error: ', error });
  }
};

export default verifyToken;

File: backend/Middlewares/VerifyToken.cs
using Data;
using Microsoft.AspNetCore.Http;
using Microsoft.EntityFrameworkCore;

namespace csharp_backend.Middlewares;

public class VerifyTokenMiddleware
{
  private readonly RequestDelegate _next;
  private readonly IServiceScopeFactory _scopeFactory;

  public VerifyTokenMiddleware(RequestDelegate next, IServiceScopeFactory scopeFactory)
  {
    _next = next;
    _scopeFactory = scopeFactory;
  }

  public async Task InvokeAsync(HttpContext context)
  {
    using (var scope = _scopeFactory.CreateScope())
    {
      var _context = scope.ServiceProvider.GetRequiredService<ApplicationDbContext>();

      var tokenHeader = context.Request.Headers["Authorization"].ToString().Split(" ")[1];
      Guid.TryParse(tokenHeader, out var token);

      var user = await _context.Users
        .Where(u => u.Token == token)
        .Select(u => new
        {
          Username = u.Username,
        })
        .FirstOrDefaultAsync();

      if (user == null)
      {
        context.Response.StatusCode = StatusCodes.Status401Unauthorized;
        await context.Response.WriteAsync("Invalid token");
        return;
      }

      context.Items["username"] = user.Username;
    }

    await _next(context);
  }
}